ledger/src/blockstore.rs

```diff
@@ -1701,69 +1773,65 @@ impl Blockstore {
     /// `duplicate_shreds`.
     ///
     /// This is intended to be used right after `shred`'s `erasure_meta`
-    /// has been created for the first time and loaded into `erasure_metas`.
-    #[allow(dead_code)]
+    /// has been created for the first time.
     fn check_forward_chained_merkle_root_consistency(
         &self,
         shred: &Shred,
+        erasure_meta: &ErasureMeta,
         just_inserted_shreds: &HashMap<ShredId, Shred>,
-        erasure_metas: &BTreeMap<ErasureSetId, WorkingEntry<ErasureMeta>>,
         merkle_root_metas: &mut HashMap<ErasureSetId, WorkingEntry<MerkleRootMeta>>,
         duplicate_shreds: &mut Vec<PossibleDuplicateShred>,
     ) -> bool {
+        debug_assert!(erasure_meta.check_coding_shred(shred));
         let slot = shred.slot();
         let erasure_set = shred.erasure_set();
-        let erasure_meta_entry = erasure_metas.get(&erasure_set).expect(
-            "Checking chained merkle root consistency on an erasure set {erasure_set:?}
-                that is not loaded in memory, programmer error",
-        );
-        let erasure_meta = erasure_meta_entry.as_ref();
-        debug_assert!(erasure_meta.check_coding_shred(shred));
 
         // If a shred from the next fec set has already been inserted, check the chaining
         let Some(next_fec_set_index) = erasure_meta.next_fec_set_index() else {
             error!("Invalid erasure meta, unable to compute next fec set index {erasure_meta:?}");
             return false;
         };
         let next_erasure_set = ErasureSetId::new(slot, next_fec_set_index);
-        let next_merkle_root_meta = match merkle_root_metas.entry(next_erasure_set) {
-            HashMapEntry::Vacant(entry) => self
-                .merkle_root_meta(next_erasure_set)
-                .unwrap()
-                .map(|meta| entry.insert(WorkingEntry::Clean(meta))),
-            HashMapEntry::Occupied(entry) => Some(entry.into_mut()),
+        let (next_shred_index, next_shred_type) = if let Some(merkle_root_meta_entry) =
+            merkle_root_metas.get(&next_erasure_set)
+        {
+            (
+                merkle_root_meta_entry.as_ref().first_received_shred_index(),
+                merkle_root_meta_entry.as_ref().first_received_shred_type(),
+            )
+        } else if let Some(merkle_root_meta) = self.merkle_root_meta(next_erasure_set).unwrap() {
+            (
+                merkle_root_meta.first_received_shred_index(),
+                merkle_root_meta.first_received_shred_type(),
+            )
+        } else {
+            // No shred from the next fec set has been received
+            return true;
```

Comment:
Might be more readable to define `next_merkle_root_meta` as below and work with that.
```rust
let Some(next_merkle_root_meta) = merkle_root_metas
    .get(&next_erasure_set)
    .map(WorkingEntry::as_ref)
    .map(Cow::Borrowed)
    .or_else(|| {
        self.merkle_root_meta(next_erasure_set)
            .unwrap()
            .map(Cow::Owned)
    })
else {
    return true; // No shred from the next fec set has been received
};
let next_shred_id = ShredId::new(
    slot,
    next_merkle_root_meta.first_received_shred_index(),
    next_merkle_root_meta.first_received_shred_type(),
);
```